### PR TITLE
Redeem invitations

### PR DESCRIPTION
**Pull request by bastjan**
## Summary

Reference: https://kb.vshn.ch/appuio-cloud/references/architecture/control-api-invitation.html#_accepting_an_invitation

This PR adds a `REDEEM` endpoint to the invitation storage and adds a controller to handle adding the user to the specified targets.

The `REDEEM` endpoint adds a snapshot of the requested permissions to the status of the invitation and marks the invitation as redeemed.

The controller watches the snapshot and adds the user that has redeemed the invitation to the requested targets.

Note: Validation of those targets is out-of-scope and will be done in another ticket.

## Checklist

- [x] PR contains a single logical change (to build a better changelog).
- [x] Categorize the PR by setting a good title and adding one of the labels:
      `bug`, `enhancement`, `documentation`, `change`, `breaking`, `dependency`
      as they show up in the changelog.

### CODE REVIEW

File: apiserver/user/invitation_redeem.go [hunk 1]
```diff
@@ -0,0 +1,133 @@
...
+func (ir *invitationRedeemer) NewConnectOptions() (runtime.Object, bool, string) {
+	// Adds the token from the path to the options under the field "token"
+	return &userv1.RedeemOptions{}, true, "token"
+}
+
+// Connect implements the REDEEM method for invitations.
+// It is used to redeem an invitation by a user.
+// The user is identified by the username in the request context.
+// The token is taken from the path.
+// If the invitation is valid, the invitation is marked as redeemed and the user and a snapshot of the users targets are stored in the status.
```
Comment (simu) [inline apiserver/user/invitation_redeem.go:47]:
```suggestion
// If the invitation is valid, the invitation is marked as redeemed and the user and a snapshot of the user's targets are stored in the status.
```

nit: "user's targets" is not very clear, given the implementation this refers to the target resources referenced in the invitation which is being redeemed.

File: apiserver/user/invitation_redeem.go [hunk 2]
```diff
@@ -0,0 +1,133 @@
...
+			return
+		}
+
+		ts := make([]userv1.TargetStatus, len(inv.Spec.TargetRefs))
+		for i, target := range inv.Spec.TargetRefs {
+			ts[i] = userv1.TargetStatus{
+				TargetRef: target,
+				Condition: metav1.Condition{
+					Type:   userv1.ConditionRedeemed,
+					Status: metav1.ConditionUnknown,
```
Comment (simu) [inline apiserver/user/invitation_redeem.go:83]: I assume there's no status value which fits better than `Unknown` (e.g. something like `Unprocessed`)

**Review by simu — APPROVED** [review]
LGTM overall, some smaller things inline

File: apis/user/v1/conversion.go [hunk 3]
```diff
@@ -0,0 +1,16 @@
+package v1
+
+import (
+	"net/url"
+
+	"k8s.io/apimachinery/pkg/conversion"
+)
+
+func Convert_url_Values_To__RedeemOptions(in, out interface{}, _ conversion.Scope) error {
```
[the remark this answers is not in the document]
Comment (bastjan) [reply apis/user/v1/conversion.go:9]: I've seen it in other code where there is a double underscore to denote "local" package.

One sample: <https://github.com/acorn-io/acorn/blob/e8a480d5d461940f29ec946caabe19c5ed2c89b0/pkg/apis/api.acorn.io/v1/scheme.go#L63>.

Upstream kubernetes seems to have a huge mix of naming options.

File: controllers/invitation_redeem_controller.go [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+		rb.Subjects, _ = ensure(rb.Subjects, newSubject(user))
+		return c.Update(ctx, &rb)
+	}
+
+	return fmt.Errorf("unsupported target %q.%q", target.APIGroup, target.Kind)
+}
+
+// ensure ensures that the given element is present in the given slice.
+// If the element is already present, the original slice is returned.
+// If the element is not present, a new slice is returned with the element appended.
```
[the remark this answers is not in the document]
Comment (bastjan) [reply controllers/invitation_redeem_controller.go:128]: I forgot in the heat. Added now 👍

File: apiserver/user/invitation_redeem.go (hunk 2 above)
Comment (bastjan) [reply apiserver/user/invitation_redeem.go:83]: Not yet. From upstream:

> In the future, we could add other intermediate conditions, e.g. ConditionDegraded.

File: apiserver/user/invitation_redeem.go (hunk 1 above)
Comment (bastjan) [reply apiserver/user/invitation_redeem.go:47]: This is misleading. I'll update the comment.

**Review by glrf — APPROVED** [review]
Took me quite a while to get what you're doing, but LGTM

File: apis/user/v1/conversion.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (simu) [reply apis/user/v1/conversion.go:9]: fair enough, I was just curious about it, I've also seen it in other projects to mark a second level of prefix/suffix separation when variables are `snake_case`.

File: apiserver/user/invitation_redeem.go (hunk 1 above)
Comment (simu) [reply apiserver/user/invitation_redeem.go]: The new version is much better, thanks :+1:

**Pull request merged**